src/SecurityAdvisoriesCheck.php

```diff
@@ -15,21 +16,63 @@ class SecurityAdvisoriesCheck extends Check
     /** @var array<string> */
     protected array $ignoredPackages = [];
 
+    protected int $retryTimes = 5;
+
+    public PackagistClient $packagistClient;
+
+    public function __construct(?PackagistClient $packagistClient = null)
+    {
+        parent::__construct();
+
+        $this->packagistClient = $packagistClient ?? new PackagistClient(new Client(), new PackagistUrlGenerator());
+    }
+
+    public function retryTimes(int $times): self
+    {
+        $this->retryTimes = $times;
+
+        return $this;
+    }
+
     public function run(): Result
     {
         $packages = $this->getInstalledPackages();
 
-        $advisories = retry(5, function () use ($packages) {
-            return $this->getAdvisories($packages);
-        }, sleepMilliseconds: 2 * 1000);
+        $packagistGatewayErrorCount = 0;
+        $lastNonGatewayPackagistError = null;
+
+        try {
+            $advisories = retry(
+                times: $this->retryTimes,
+                callback: function () use ($packages) {
+                    return $this->getAdvisories($packages);
+                },
+                sleepMilliseconds: 2 * 1000,
+                when: function ($e) use (&$packagistGatewayErrorCount, &$lastNonGatewayPackagistError) {
+                    if ($e instanceof ServerException && in_array($e->getCode(), [502, 503, 504])) {
```

Comment:
Rename `$e` to `$exception`

Also refactor this whole condition this it's own function. Avoid using `$$` by using early returns in that function. 

That will make it more readable.